### PR TITLE
[PPP-4772] RCE injection via connection's JNDI database name

### PR DESCRIPTION
**Pull request by dcleao**
Issue: https://hv-eng.atlassian.net/browse/PPP-4772

Merge with:
- https://github.com/pentaho/pentaho-platform-ee/pull/1773

/cc @pentaho/hoth

### CODE REVIEW

**Comment by buildguy** [comment]
### :x: Build failed in 36m 26s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions,core
```
---
:ok_hand: All tests passed!
 
Tests run: **570**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/platform/job/PR-5674/1/testReport)

 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code -2
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 19m 57s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions,core
```
---
:ok_hand: All tests passed!
 
Tests run: **570**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/platform/job/PR-5674/2/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 24m 18s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions,core
```
---
:ok_hand: All tests passed!
 
Tests run: **570**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/platform/job/PR-5674/3/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Review by singletonc — APPROVED** [review]

File: core/src/main/java/org/pentaho/platform/engine/services/connection/datasource/dbcp/PooledDatasourceHelper.java [hunk 1]
```diff
@@ -349,60 +348,172 @@ private static void initDriverClass( DriverManagerDataSource driverManagerDataSo
...
-      if ( firstNe != null ) {
-        throw new DBDatasourceServiceException( firstNe );
-      }
-      throw new DBDatasourceServiceException( dsName );
-    } catch ( NamingException ne ) {
-      throw new DBDatasourceServiceException( ne );
+    }
+
+    // Always non-null.
+    throw firstNe;
```
Comment (singletonc) [inline core/src/main/java/org/pentaho/platform/engine/services/connection/datasource/dbcp/PooledDatasourceHelper.java:448]: Sonar is complaining about this, not a blocker

File: core/src/main/java/org/pentaho/platform/engine/services/connection/datasource/dbcp/PooledDatasourceHelper.java [hunk 2]
```diff
@@ -349,60 +348,172 @@ private static void initDriverClass( DriverManagerDataSource driverManagerDataSo
...
+   * @return The data source; never {@code null}.
+   * @throws DBDatasourceServiceException If the data source name is not from a valid namespace, is not defined, or the
+   *                                      name resolves to a resource which does not implement {@link DataSource}.
+   */
+  public static DataSource getJndiDataSource( Context context, String dsName, List<String> allowedJndiSchemes )
+    throws DBDatasourceServiceException {
+
+    String[] candidateNames = {
       // First, try what they ask for...
+      dsName,
```
Comment (NJtwentyone) [inline core/src/main/java/org/pentaho/platform/engine/services/connection/datasource/dbcp/PooledDatasourceHelper.java:397]: good java doc comments list of candidatesNames prefixes

File: core/src/test/java/org/pentaho/platform/engine/services/connection/datasource/dbcp/PooledDatasourceHelperTest.java [hunk 3]
```diff
@@ -347,6 +353,174 @@ public void testConnectionFactory_MicrosoftSQL() {
...
+    Tomcat1JndiScenario scenario = new Tomcat1JndiScenario();
+
+    scenario.getJndiDataSource( scenario.DATA_SOURCE_3_RMI_FULL_NAME );
+
+    verify( scenario.context, never() ).lookup( anyString() );
+  }
+
+  @Test( expected = DBDatasourceServiceException.class )
+  public void testGetJndiDataSourceThrowsGivenNameWithDisallowedJavaScheme() throws Exception {
+    // This tests that it is possible to disable the java scheme via configuration!
```
Comment (NJtwentyone) [inline core/src/test/java/org/pentaho/platform/engine/services/connection/datasource/dbcp/PooledDatasourceHelperTest.java:514]: good tests

File: core/src/main/java/org/pentaho/platform/engine/services/connection/datasource/dbcp/PooledDatasourceHelper.java [hunk 4]
```diff
@@ -349,60 +348,172 @@ private static void initDriverClass( DriverManagerDataSource driverManagerDataSo
...
+   * @param context            The naming context.
+   * @param dsName             The data source name.
+   * @param allowedJndiSchemes The list of allowed JNDI schemes of valid data source names.
+   * @return The data source resource; never {@code null}.
+   * @throws NamingException    If the given data source name is not from a valid namespace, or if it does not exist
+   *                            in the given context.
+   * @throws ClassCastException If the given data source name resolves is that of a resource which does not implement
+   *                            {@link DataSource}.
+   */
+  private static DataSource lookupJndiDataSourceByName( Context context,
```
Comment (NJtwentyone) [inline core/src/main/java/org/pentaho/platform/engine/services/connection/datasource/dbcp/PooledDatasourceHelper.java:463]: code is really readable and maintainable now.

**Review by NJtwentyone — APPROVED** [review]
good pr

**Pull request merged**